bpf/lib/overloadable_xdp.h

```diff
@@ -167,12 +177,25 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
 		   __u32 node_id __maybe_unused,
 		   __u32 seclabel __maybe_unused,
 		   __u32 dstid __maybe_unused,
-		   __u32 vni __maybe_unused, int *ifindex __maybe_unused)
+		   __u32 vni __maybe_unused,
+		   void *opt __maybe_unused,
+		   __u32 opt_len __maybe_unused,
+		   bool is_ipv6 __maybe_unused,
+		   int *ifindex __maybe_unused)
 {
 	ctx_store_meta(ctx, CB_ENCAP_NODEID, bpf_ntohl(node_id));
 	ctx_store_meta(ctx, CB_ENCAP_SECLABEL, seclabel);
 	ctx_store_meta(ctx, CB_ENCAP_DSTID, dstid);
 	ctx_set_xfer(ctx, XFER_PKT_ENCAP);
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+	if (opt && !is_ipv6) {
+		__be32 *addr = opt + sizeof(struct geneve_opt_hdr);
+		__be16 *port = opt + sizeof(struct geneve_opt_hdr) + sizeof(__be32);
+
+		ctx_store_meta(ctx, CB_ENCAP_PORT, *port);
+		ctx_store_meta(ctx, CB_ENCAP_ADDR, *addr);
+	}
+#endif
```

Comment:
Should we return a DROP reason if it's IPv6, just to be clear that it is unhandled currently?